File: tfcc_code_generator/frontend/onnx/converter/nn/gather.py
# Copyright 2021 Wechat Group, Tencent
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import onnx
import ir.node
import ir.framework
import frontend.onnx.common
from ..converter import Converter


class Gather(Converter):
    def __call__(
        self,
        node_proto: onnx.NodeProto,
        graph_proto: onnx.GraphProto,
        graph: ir.framework.Graph,
    ):
        if not self.accept(node_proto):
            return False

        attributes = self.get_attributes(node_proto)
        axis = 0
        if "axis" in attributes:
            axis = attributes["axis"]

        input_symbol = graph.get_symbol(node_proto.input[0])
        indices_symbol = graph.get_symbol(node_proto.input[1])
        while axis < 0:
            axis = axis + len(input_symbol.shape)

        if input_symbol.is_tensor() and indices_symbol.is_tensor():
            graph.append_node(
                ir.node.nn.Gather(
                    node_proto.name,
                    graph,
                    node_proto.input,
                    node_proto.output,
                    axis=axis,
                )
            )
            return True
        elif (
            input_symbol.is_tensor()
            and indices_symbol.is_value()
            and len(input_symbol.shape) == 1
        ):
            toVectorOutputName = graph.context.create_symbol_name(node_proto.output[0])
            graph.append_node(
                ir.node.base.ToVector(
                    node_proto.name + ":0",
                    graph,
                    [node_proto.input[0]],
                    [toVectorOutputName],
                )
            )
            graph.append_node(
                ir.node.base.At(
                    node_proto.name + ":1",
                    graph,
                    [toVectorOutputName, node_proto.input[1]],
                    node_proto.output,
                )
            )
            return True
        elif (
            input_symbol.is_tensor()
            and indices_symbol.is_value()
            and len(input_symbol.shape) > 1
        ):
            sliceOutputName = graph.context.create_symbol_name(node_proto.output[0])
            graph.append_node(
                ir.node.base.SliceV1(
                    node_proto.name + ":0",
                    graph,
                    node_proto.input,
                    [sliceOutputName],
                    axis=axis,
                    length=1,
                )
            )
            graph.append_node(
                ir.node.base.Squeeze(
                    node_proto.name + ":1",
                    graph,
                    [sliceOutputName],
                    node_proto.output,
                    axes=[axis],
                )
            )
            return True
        else:
            return False

    @property
    def accept_versions(self) -> set:
        # TODO Support negative indices
        return set([1, 11])

File: tfcc_runtime/operations/nn/gather.cpp
// Copyright 2021 Wechat Group, Tencent
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "gather.h"

#include <algorithm>

#include "tfcc_runtime/framework/graph.h"
#include "tfcc_runtime/proto/operations/nn.pb.h"
#include "tfcc_runtime/utils/commutils.h"

namespace tfcc {
namespace runtime {
namespace nn {

template <class T, class IDX>
static const char* _wrapper_gather(
    const tfcc::Tensor<T>* a, const tfcc::Tensor<IDX>* indices, size_t axis,
    tfcc::Variable<T>* result) noexcept {
  try {
    if (axis != 0) {
      return "Gather: invalid axis";
    }
    *result = tfcc::nn::gather(*a, *indices);
  } catch (std::exception& e) {
    thread_local std::string reason = e.what();
    return reason.c_str();
  }
  return nullptr;
}

template <tfcc::runtime::common::DataType DTYPE, tfcc::runtime::common::DataType IDX_DTYPE>
std::string Gather<DTYPE, IDX_DTYPE>::getOperationTypeUrl() const {
  tfcc::runtime::operations::nn::Gather operation;
  return get_protobuf_type_url(operation);
}

template <tfcc::runtime::common::DataType DTYPE, tfcc::runtime::common::DataType IDX_DTYPE>
std::set<unsigned> Gather<DTYPE, IDX_DTYPE>::getOperationVersions() const {
  std::set<unsigned> versions = {tfcc::runtime::operations::nn::Gather::VERSION_1};
  return versions;
}

template <tfcc::runtime::common::DataType DTYPE, tfcc::runtime::common::DataType IDX_DTYPE>
std::unique_ptr<OperationResource> Gather<DTYPE, IDX_DTYPE>::process(
    const tfcc::runtime::model::Node& node, Graph& graph) const {
  if (node.inputs_size() != 2 || node.outputs_size() != 1) {
    return nullptr;
  }
  if (!node.operation().Is<tfcc::runtime::operations::nn::Gather>()) {
    return nullptr;
  }
  auto& symbolManager = getSymbolManager(graph);
  if (symbolManager.getSymbolInfo(node.outputs(0)).dtype != DTYPE ||
      symbolManager.getSymbolInfo(node.inputs(0)).dtype != DTYPE) {
    return nullptr;
  }
  const SymbolInfo& indicesSymbolInfo = symbolManager.getSymbolInfo(node.inputs(1));
  if (indicesSymbolInfo.dtype != IDX_DTYPE) {
    return nullptr;
  }

  std::unique_ptr<OperationResource> resource(new OperationResource);
  tfcc::runtime::operations::nn::Gather operation;
  node.operation().UnpackTo(&operation);

  auto& jit = getJIT(graph);

  const tfcc::Tensor<T>* inputSymbol = symbolManager.getTensor(node.inputs(0), T());
  const tfcc::Tensor<IDX>* indicesSymbol = symbolManager.getTensor(node.inputs(1), IDX());
  tfcc::Variable<T>* outputSymbol = symbolManager.getVariable(node.outputs(0), T());

  callFunction(
      jit, _wrapper_gather<T, IDX>, inputSymbol, indicesSymbol, operation.axis(), outputSymbol);

  return resource;
}

std::vector<std::unique_ptr<Operation>> get_gather_operations() {
#define DEFINE_FUNC(dtype)                                                             \
  operations.emplace_back(                                                             \
      std::unique_ptr<Operation>(new Gather<dtype, tfcc::runtime::common::INT32>()));  \
  operations.emplace_back(                                                             \
      std::unique_ptr<Operation>(new Gather<dtype, tfcc::runtime::common::UINT32>())); \
  operations.emplace_back(                                                             \
      std::unique_ptr<Operation>(new Gather<dtype, tfcc::runtime::common::INT64>()));  \
  operations.emplace_back(                                                             \
      std::unique_ptr<Operation>(new Gather<dtype, tfcc::runtime::common::UINT64>()));

  std::vector<std::unique_ptr<Operation>> operations;
  TFCC_RUNTIME_FOR_ALL_DATA_TYPE(DEFINE_FUNC);
  return operations;
}

}  // namespace nn
}  // namespace runtime
}  // namespace tfcc
